gradle-jdks-settings/src/main/java/com/palantir/gradle/jdks/settings/PatchToolchainJdkPlugin.java

```diff
@@ -0,0 +1,203 @@
+        if (!Files.exists(gradleJdksLocalDirectory)) {
+            logger.debug("Not setting the Gradle JDK properties because gradle/jdks directory doesn't exist. Please run"
+                    + " ./gradlew setupJdks to set up the JDKs.");
+            return;
+        }
+        ProviderFactory providerFactory =
+                ((DefaultSettings) settings).getServices().get(ProviderFactory.class);
+        if (!(providerFactory instanceof DefaultProviderFactory)) {
+            throw new RuntimeException(String.format(
+                    "Expected providerFactory to be of type '%s' but was '%s'.",
+                    ProviderFactory.class.getCanonicalName(),
+                    providerFactory.getClass().getCanonicalName()));
+        }
+        DefaultProviderFactory defaultProviderFactory = (DefaultProviderFactory) providerFactory;
+        try {
+            Field valueSourceProviderFactory =
+                    DefaultProviderFactory.class.getDeclaredField("valueSourceProviderFactory");
+            valueSourceProviderFactory.setAccessible(true);
+
+            DefaultValueSourceProviderFactory defaultValueSourceProviderFactory =
+                    (DefaultValueSourceProviderFactory) valueSourceProviderFactory.get(defaultProviderFactory);
+            Field field = DefaultValueSourceProviderFactory.class.getDeclaredField("gradleProperties");
+            field.setAccessible(true);
+            GradleProperties originalGradleProperties = (GradleProperties) field.get(defaultValueSourceProviderFactory);
+            GradleProperties ourGradleProperties = (GradleProperties) Proxy.newProxyInstance(
+                    GradleProperties.class.getClassLoader(),
+                    new Class[] {GradleProperties.class},
+                    new GradlePropertiesInvocationHandler(gradleJdksLocalDirectory, originalGradleProperties));
+            field.set(defaultValueSourceProviderFactory, ourGradleProperties);
+        } catch (NoSuchFieldException | IllegalAccessException e) {
+            throw new RuntimeException("Failed to update the Gradle JDK properties using reflection", e);
+        }
+    }
+
+    static Path getToolchainInstallationDir() {
+        return Path.of(Optional.ofNullable(System.getenv("GRADLE_USER_HOME"))
+                        .orElseGet(() -> System.getProperty("user.home") + "/.gradle"))
+                .resolve("gradle-jdks");
+    }
+
+    // keep in sync with com.palantir.gradle.jdks.JdksPlugin#isGradleJdkSetupEnabled(org.gradle.api.Project)
+    public static boolean isGradleJdkSetupEnabled(Path projectDir) {
+        if (GradleVersion.current().compareTo(GradleVersion.version("7.6")) < 0) {
+            throw new RuntimeException("Failed to apply com.palantir.jdks.settings plugin. Gradle JDK setup is enabled"
+                    + " (palantir.jdk.setup.enabled=true) but the version of Gradle is less than 7.6. Please"
+                    + " upgrade to Gradle 7.6 or higher to use this functionality.");
```

Comment:
Just for my understanding (I think it's fine to not support <7.6), is this due to some internal code change around 7.6?

Reply:
I tried testing with different Gradle Versions (https://github.com/palantir/gradle-jdks/pull/360/files#diff-98ba650ac01c560cdf5ffae45e4a36b0f4786c3956183b2d9cc1d9759b3061ceR87) but anything below Gradle `v7.6` it would fail with `Unsupported class file major version`.